File: services/anomaly-detector/health.go
package main

import (
	"context"
	"encoding/json"
	"log"
	"net/http"
	"time"

	"github.com/prometheus/client_golang/prometheus/promhttp"
)

// startHealthServer exposes /live (process is up — never fails on a
// dependency), /ready (this instance can currently do useful work — the
// device/sensor catalog it needs for detection is Postgres-backed, so a
// Postgres outage means it genuinely isn't ready, and an open Kafka
// breaker means it can detect but not actually publish results), /forests,
// and /metrics.
func startHealthServer(port string, cat *catalog, kio *kafkaIO, registry *forestRegistry) {
	mux := http.NewServeMux()

	mux.HandleFunc("/live", func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte("ok"))
	})

	mux.HandleFunc("/ready", func(w http.ResponseWriter, r *http.Request) {
		ctx, cancel := context.WithTimeout(r.Context(), 2*time.Second)
		defer cancel()
		postgresOK := cat.ping(ctx) == nil
		breakerState := kio.breakerState()
		ready := postgresOK && breakerState != "OPEN"

		w.Header().Set("Content-Type", "application/json")
		if !ready {
			w.WriteHeader(http.StatusServiceUnavailable)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{
			"ready":                 ready,
			"postgres_connected":    postgresOK,
			"kafka_circuit_breaker": breakerState,
		})
	})

	mux.HandleFunc("/forests", func(w http.ResponseWriter, r *http.Request) {
		registry.mu.RLock()
		defer registry.mu.RUnlock()
		types := make([]string, 0, len(registry.forests))
		for mt := range registry.forests {
			types = append(types, mt)
		}
		w.Header().Set("Content-Type", "application/json")
		_ = json.NewEncoder(w).Encode(map[string]any{"trained_machine_types": types})
	})

	mux.Handle("/metrics", promhttp.Handler())

	go func() {
		// ListenAndServe only returns once the listener stops; nothing in
		// this service ever calls Shutdown/Close on it, so any return here
		// is unexpected (most likely a bind failure, e.g. the port already
		// in use) and worth surfacing loudly rather than leaving /live and
		// /ready silently unreachable for the rest of the process's life.
		if err := http.ListenAndServe(":"+port, mux); err != nil { //nolint:gosec // internal-only health/metrics endpoint
			log.Printf("anomaly-detector: health/metrics server on :%s stopped: %v", port, err)
		}
	}()
}

File: services/anomaly-detector/health.py
"""HTTP health/metrics server, the Python port of health.go. Exposes /live
(process is up — never fails on a dependency), /ready (this instance can
currently do useful work — the device/sensor catalog it needs for
detection is Postgres-backed, so a Postgres outage means it genuinely
isn't ready, and an open Kafka breaker means it can detect but not
actually publish results), /forests, and /metrics.
"""

from __future__ import annotations

import json
import logging
import threading
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

from prometheus_client import CONTENT_TYPE_LATEST, generate_latest

from catalog import Catalog
from forestregistry import ForestRegistry
from kafka_io import KafkaIO

_logger = logging.getLogger("anomaly-detector")


def _make_handler(cat: Catalog, kio: KafkaIO, registry: ForestRegistry) -> type[BaseHTTPRequestHandler]:
    class Handler(BaseHTTPRequestHandler):
        def log_message(self, fmt: str, *args) -> None:  # noqa: A002 - silence stdlib's default access log
            pass

        def do_GET(self) -> None:
            if self.path == "/live":
                self.send_response(200)
                self.end_headers()
                self.wfile.write(b"ok")
                return

            if self.path == "/ready":
                postgres_ok = cat.ping()
                breaker_state = kio.breaker_state()
                ready = postgres_ok and breaker_state != "OPEN"
                body = json.dumps(
                    {"ready": ready, "postgres_connected": postgres_ok, "kafka_circuit_breaker": breaker_state}
                ).encode("utf-8")
                self.send_response(200 if ready else 503)
                self.send_header("Content-Type", "application/json")
                self.end_headers()
                self.wfile.write(body)
                return

            if self.path == "/forests":
                body = json.dumps({"trained_machine_types": registry.trained_machine_types()}).encode("utf-8")
                self.send_response(200)
                self.send_header("Content-Type", "application/json")
                self.end_headers()
                self.wfile.write(body)
                return

            if self.path == "/metrics":
                body = generate_latest()
                self.send_response(200)
                self.send_header("Content-Type", CONTENT_TYPE_LATEST)
                self.end_headers()
                self.wfile.write(body)
                return

            self.send_response(404)
            self.end_headers()

    return Handler


def start_health_server(port: str, cat: Catalog, kio: KafkaIO, registry: ForestRegistry) -> ThreadingHTTPServer | None:
    """Starts the health/metrics server in a background thread. A bind
    failure (e.g. the port already in use) is logged loudly rather than
    silently leaving /live and /ready unreachable for the process's whole
    lifetime — the exact fix a pre-GitHub audit made to health.go, ported
    here rather than reintroducing the original bug."""
    try:
        server = ThreadingHTTPServer(("", int(port)), _make_handler(cat, kio, registry))
    except OSError as exc:
        _logger.error("anomaly-detector: health/metrics server on :%s stopped: %s", port, exc)
        return None
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()
    return server
